### ISSUE TITLE
Issue with Ionic/Capacitor when using a production build.

### ISSUE THREAD

**Issue by mxdmedia** [briebug/ngrx-auto-entity#81]
I've got a pretty big issue that just cropped up- auto-entity's facades/selectors are causing an issue with Ionic/Angular production builds (at least on iOS, I don't have an android device to test).  Basically, any time a facade selector is called, a javascript error occurs, and the selector doesn't return its results.  This only happens under the following conditions:

1. A facade selector (e.g.-  myFacade.all$) is triggered.
2. The build is a production build (e.g.- `ionic build --prod`)
3. The production build app is run on an iOS device.

If I generate a debug build, everything runs as expected on the device.  If I generate a production build and run it in a browser, everything runs as expected.  If I use vanilla ngrx selectors, everything runs as expected.

What makes this even more difficult- the ios console wont output source map errors as far as I can tell.  Thus, the entirety of the error I am seeing is this:
```
ERROR {"line":1,"column":442539,"sourceURL":"capacitor://localhost/main-es2015.24b639144b74affc753b.js"}
```
Looking at the javascript file in question, column 442539 is:
```
o=(e=>({selectAll:Object(r.y)(e,e=>e.ids.map(t=>e.entities[t]))
```
Not a lot to go on.  Do you have any thoughts on what would cause errors in this case?  I'll ping the Capacitor/Ionic people, and see if there is a better way to debug a production compiled app on the device, and will update this issue as I learn more.

Note- there is a very simplified repo exhibiting the issue in the Capacitor issue I filed (link below)

**Comment by jrista** [comment]
Thanks for the report. We'll look into it.

In the interim, for debugging the prod app. You could temporarily go into angular.json, and adjust the production build target's settings so it does not minify, continues to generate source maps, etc. That might help you zero in on the error more closely.

**Comment by mxdmedia** [comment]
Thanks for the tips on adjusting the build.  I've tried a few things and have learned the following:

1. I dont think the webview that capacitor/ionic uses on iOS supports source maps, as even when I enable them (and see they are created/in my app bundle) the error is still the same not-so-helpful error message above.  I'm hopeful that the capacitor folks might be able to shed some light on getting better error messages.
2. It appears that whatever the `--optimize` flag (NOT `--optimize-build`) does, is causing the issue.  There doesn't seem to be a lot of info about what `--optimize` does (the angular docs just say it... optimizes), but it looks like it performs minification at the very least (when turned off, the code is readable again).  Once I turn `optimize` off, the app runs without issue, even if I have `aot` and `optimize-build` turned on.  So in the short term, this seems to be a good option.  Performance is good no matter the build, but with debug, the bundle size is quite large (~12m).  The non-working version with `--optimize` on is 3.2m, and the working production version with `--optimize` off is 6.4m, which while a bit larger than I'd like (we do OTA push updates), is still much better than the debug size.

**Comment by jrista** [comment]
Which WebView is capacitor/ionic using? I generally use cordova with ionic (I prefer to stay out of native code :P), and I do seem to get source mapped code in the browser. I am not sure if capacitor projects use the same webview though.

Thanks for the detail about the --optimize flag. I wonder if something is getting tree-shaken out that should not be... We'll look into it. It would be interesting to see if you can get source maps working with --optimize enabled...that would at least allow us to delve into what code might be missing, if that is indeed what's happening.

**Comment by mxdmedia** [comment]
Tomorrow I'll ask about getting source mapped output on the Capacitor slack channel to see if I can make some headway there.

As to the webview, Capacitor is using WKWebView (as opposed to the older, deprecated UIWebView).

**Comment by mxdmedia** [comment]
I figured out how to get more information, and have some sample code here:  https://github.com/mxdmedia/ionic-ngrx-ae-issue

To test/see the error (you'll need a mac of course):

```
npm install
ionic build --prod
npx cap copy
npx cap open ios
```

On your ios device go to **settings -> safari -> advanced** and make sure Web Inspector is turned on.

Connect said device to your mac.  Then in xcode, build/run on your device.  While xcode is building/installing, fire up Safari on your mac.  After enabling the developer menu, you should see your device name in the Developer menu.  Select it, then wait for the app to start running on your device.  Once it is running, select "localhost" under the app name.  This will open a developer console.  Make sure breakpoints are disabled (an uncaught exception breakpoint wont be triggered with this error).  The error occurs right when the app is launched, so you'll need to reload.  In the web inspector/console window, there is a reload button in the upper-left corner (next to the "download web archive" button).  Click this, and it will reload the app, and in the console tab, you'll see an error appear.  The error is:

```
TypeError: undefined is not an object (evaluating 'e.ids') — briebug-ngx-auto-entity.js.pre-build-optimizer.js:1398
```
In my larger code base (which I cant make available unfortunately), I get that error, along with:
```
TypeError: undefined is not an object (evaluating 'e.entities') — briebug-ngx-auto-entity.js.pre-build-optimizer.js:1407
```

I hope this is helpful.

**Comment by jrista** [comment]
That sounds like the core state we create for each entity. Every entity, following the pattern set forth by @ngrx/entity, has an entities object (associative array or map, as it is used) and an ids array. It sounds like something about the optimization process is preventing the initial state from being initialized properly... 

I'll take a look at your example soon here.

**Comment by mxdmedia** [comment]
I noticed something that might make this a bit easier to work on.  It turns out, the same error is thrown even in the browser with `ionic build --prod`.  I was trying `ionic serve --prod`, which I thought was doing the same production build, but it must be doing something different.  If you test using the `ionic serve` command (with `--prod`) there is no error, and it all works fine.  On the other hand, if you do `ionic build --prod`, then serve up the `www` directory that is generated (using either a node web server, or something like nginx), you will see the same errors regarding `.ids` and `.entities` in the browser.

**Comment by jrista** [comment]
I am checking out your reproduction repo. I'm generating two builds, one with --optimize and one without. I'll then be comparing the outputs to see what is different. I wonder if somehow something is getting tree-shaken out of the ngrx-ae library code...

**Comment by jrista** [comment]
Well, --optimize is really where all the optimization happens. Without this flag, the compiled .js files are still prettily formatted, still contain original identifiers, etc. With the flag, everything gets minified and uglified and all that. 

From what I can tell, there is only one place in the code where e.ids and e.entities is found in the optimized build. That is in the pre-fabbed selectors that come with ngrx-auto-entity. Right now, those selectors do not have any null or undefined checks. I wouldn't have thought I would have needed them, as `e` in this context is apparently the root state for the entity itself... I am not sure how or why the root state of the entity would be undefined. However I can add some falsy checks and see if that solves the problem.

**Comment by jrista** [comment]
@mxdmedia I think seeing the full stack traces you are getting would also be helpful. I am trying to get some of my own, but it would be helpful to have yours as well.

**Comment by jrista** [comment]
@mxdmedia I seem to be having trouble reproducing the error. I've used the `ionic build --prod` approach along with `http-server` to serve up the code in ./www. So far, I have not yet encountered the error...in Safari, or in any Chromium browser or in FireFox...

If you can share your stack traces, that might help me figure out where the issue is being initiated, at the very least.

**Comment by mxdmedia** [comment]
@jrista-

I hadn't tried `ionic build --prod` with serving, with the sample code (it was using my much larger codebase that was causing the issues with `ionic build --prod`- perhaps it was an unrelated issue).

Here is the stack trace I am seeing from my iPhone via the safari debugger (using the code you have)  I am not terribly familiar with dumping a stacktrace from the Safari debugging tools, so if you know a way to get more details, I am all ears:

```
TypeError: undefined is not an object (evaluating 'e.ids')
(anonymous function) — briebug-ngrx-auto-entity.js.pre-build-optimizer.js:1398
memoized — store.js.pre-build-optimizer.js:861
memoized — store.js.pre-build-optimizer.js:861
_next — map.js.pre-build-optimizer.js:29
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_subscribe — ReplaySubject.js.pre-build-optimizer.js:57
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:27
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:27
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
ngOnInit — home.page.ts:32
(anonymous function) — core.js.pre-build-optimizer.js:27788
jc — core.js.pre-build-optimizer.js:39265
(anonymous function) — home.page.ngfactory.js.pre-build-optimizer.js:27:284
rc — core.js.pre-build-optimizer.js:38376
detectChanges — core.js.pre-build-optimizer.js:27092
(anonymous function) — fesm5.js.pre-build-optimizer.js:3601
(anonymous function) — fesm5.js.pre-build-optimizer.js:4246
activateRoutes — router.js.pre-build-optimizer.js:3132
(anonymous function) — router.js.pre-build-optimizer.js:3069
forEach
activateChildRoutes — router.js.pre-build-optimizer.js:3068
activateRoutes — router.js.pre-build-optimizer.js:3139
(anonymous function) — router.js.pre-build-optimizer.js:3069
forEach
activateChildRoutes — router.js.pre-build-optimizer.js:3068
activate — router.js.pre-build-optimizer.js:2927
(anonymous function) — router.js.pre-build-optimizer.js:2900
_next — map.js.pre-build-optimizer.js:29
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — switchMap.js.pre-build-optimizer.js:67
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:3
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — switchMap.js.pre-build-optimizer.js:46
_next — switchMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — switchMap.js.pre-build-optimizer.js:67
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — mergeMap.js.pre-build-optimizer.js:69
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_complete — takeLast.js.pre-build-optimizer.js:52
complete — Subscriber.js.pre-build-optimizer.js:61
_complete — Subscriber.js.pre-build-optimizer.js:79
complete — Subscriber.js.pre-build-optimizer.js:61
_complete — mergeMap.js.pre-build-optimizer.js:64
complete — Subscriber.js.pre-build-optimizer.js:61
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:5
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — mergeMap.js.pre-build-optimizer.js:59
_tryNext — mergeMap.js.pre-build-optimizer.js:53
_next — mergeMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:3
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — switchMap.js.pre-build-optimizer.js:46
_next — switchMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
_next — filter.js.pre-build-optimizer.js:33
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — mergeMap.js.pre-build-optimizer.js:69
_next — InnerSubscriber.js.pre-build-optimizer.js:11
```

**Comment by mxdmedia** [comment]
Sometimes one doesn't see the forest for the trees.  Right clicking the error has a copy all option.  Here it is, with some better formatting an a bit more info.  Sorry about polluting the thread a bit with my last paste.
```
[Error] ERROR
TypeError: undefined is not an object (evaluating 'e.ids')
(anonymous function) — briebug-ngrx-auto-entity.js.pre-build-optimizer.js:1398
memoized — store.js.pre-build-optimizer.js:861
memoized — store.js.pre-build-optimizer.js:861
_next — map.js.pre-build-optimizer.js:29
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_subscribe — ReplaySubject.js.pre-build-optimizer.js:57
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:27
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:27
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
ngOnInit — home.page.ts:32
(anonymous function) — core.js.pre-build-optimizer.js:27788
jc — core.js.pre-build-optimizer.js:39265
(anonymous function) — home.page.ngfactory.js.pre-build-optimizer.js:27:284
rc — core.js.pre-build-optimizer.js:38376
detectChanges — core.js.pre-build-optimizer.js:27092
(anonymous function) — fesm5.js.pre-build-optimizer.js:3601
(anonymous function) — fesm5.js.pre-build-optimizer.js:4246
activateRoutes — router.js.pre-build-optimizer.js:3132
(anonymous function) — router.js.pre-build-optimizer.js:3069
forEach
activateChildRoutes — router.js.pre-build-optimizer.js:3068
activateRoutes — router.js.pre-build-optimizer.js:3139
(anonymous function) — router.js.pre-build-optimizer.js:3069
forEach
activateChildRoutes — router.js.pre-build-optimizer.js:3068
activate — router.js.pre-build-optimizer.js:2927
(anonymous function) — router.js.pre-build-optimizer.js:2900
_next — map.js.pre-build-optimizer.js:29
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — switchMap.js.pre-build-optimizer.js:67
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:3
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — switchMap.js.pre-build-optimizer.js:46
_next — switchMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — switchMap.js.pre-build-optimizer.js:67
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — mergeMap.js.pre-build-optimizer.js:69
_next — InnerSubscriber.js.pre-build-optimizer.js:11
next — Subscriber.js.pre-build-optimizer.js:49
_next — map.js.pre-build-optimizer.js:35
next — Subscriber.js.pre-build-optimizer.js:49
_complete — takeLast.js.pre-build-optimizer.js:52
complete — Subscriber.js.pre-build-optimizer.js:61
_complete — Subscriber.js.pre-build-optimizer.js:79
complete — Subscriber.js.pre-build-optimizer.js:61
_complete — mergeMap.js.pre-build-optimizer.js:64
complete — Subscriber.js.pre-build-optimizer.js:61
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:5
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — mergeMap.js.pre-build-optimizer.js:59
_tryNext — mergeMap.js.pre-build-optimizer.js:53
_next — mergeMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
(anonymous function) — subscribeToArray.js.pre-build-optimizer.js:3
_trySubscribe — Observable.js.pre-build-optimizer.js:42
subscribe — Observable.js.pre-build-optimizer.js:28
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
subscribe — Observable.js.pre-build-optimizer.js:23
_innerSub — switchMap.js.pre-build-optimizer.js:46
_next — switchMap.js.pre-build-optimizer.js:36
next — Subscriber.js.pre-build-optimizer.js:49
_next — filter.js.pre-build-optimizer.js:33
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
_next — tap.js.pre-build-optimizer.js:46
next — Subscriber.js.pre-build-optimizer.js:49
notifyNext — mergeMap.js.pre-build-optimizer.js:69
_next — InnerSubscriber.js.pre-build-optimizer.js:11
	capacitorConsole (user-script:2:78)
	fn (main-es2015.13dd38d54f3908ea9612.js:1:37970)
	handleError (main-es2015.13dd38d54f3908ea9612.js:1:38131)
	next (main-es2015.13dd38d54f3908ea9612.js:1:101025)
	(anonymous function) (main-es2015.13dd38d54f3908ea9612.js:1:91506)
	__tryOrUnsub (main-es2015.13dd38d54f3908ea9612.js:1:15235)
	next (main-es2015.13dd38d54f3908ea9612.js:1:14464)
	_next (main-es2015.13dd38d54f3908ea9612.js:1:13627)
	next (main-es2015.13dd38d54f3908ea9612.js:1:13402)
	next (main-es2015.13dd38d54f3908ea9612.js:1:195031)
	emit (main-es2015.13dd38d54f3908ea9612.js:1:91377)
	run (polyfills-es2015.2be0e5ac9028b106c374.js:1:1993)
	onHandleError (main-es2015.13dd38d54f3908ea9612.js:1:95521)
	runTask (polyfills-es2015.2be0e5ac9028b106c374.js:1:2667)
	invokeTask (polyfills-es2015.2be0e5ac9028b106c374.js:1:8117)
	(anonymous function) (polyfills-es2015.2be0e5ac9028b106c374.js:1:32067)
```

**Comment by jrista** [comment]
@mxdmedia Ok, looks like the first line of your example code that is related is the subscribe line in home.page.ts. 

I am still not encountering the error on my end. So can I have to do a bit more for me? Can you scatter console.log statements throughout your demo code? Kind of like this:

```typescript
import { Sample } from './../models/sample.model';
import { buildState, IEntityState } from '@briebug/ngrx-auto-entity';

console.log('sample.state: enter');

export const {
  initialState: sampleInitialState,
  selectors: sampleSelectors,
  facade: SampleFacadeBase
} = buildState(Sample);

console.log('sample.state: post buildStatE()');

export function sampleReducer(
  state = sampleInitialState
): IEntityState<Sample> {
  return state;
}

console.log('sample.state: post reducer');
```

Do this throughout all the various state related files, the routing files, module and component files. I just need to figure out in what order things are happening... You might also, just for kicks, try to change the order of the imports in your main app.module. You have the routing module imported before the state module...as strange as it sounds, I recently ran into an issue with the order in which I imported modules and routing in an Ionic 4/Cordova app. My issue was not related to auto-entity, but when I rearranged how I was importing, and put the rou

ter module import last, it resolved my issue. The issue I was having honestly made no sense, and I don't know how it occurred in the first place...and even rearranging the module import order since back to how it was has not caused the issue to reoccur (YET!) Anyway, it was a very odd issue that had me spinning my wheels for a couple of days with Ionic 4 recently, and I wonder if you may be running into some kind of similar issue.

**Comment by jrista** [comment]
I would also be curious to know, if rearranging the imports does not help, whether changing the home route from lazy loaded to static would change things. I have also recently had issues trying to get multiple lazy loaded routes working with Ionic 4. Again, a very odd an unexplainable issue so far, if I lazy load page modules, things work fine, the third and my app hangs. Doesn't seem to matter which module is lazy loaded third in the routes, all that seems to matter in my case is that three modules are lazy loaded.

**Comment by mxdmedia** [comment]
@jrista, I can definitely add the console logging, try re-ordering the imports, and see how static vs. lazy loading changes things.  Unfortunately, I am running a bit short on time today, but will get to it either tomorrow or Monday.  Thanks again for your time and effort- it is much appreciated!

**Comment by jrista** [comment]
@mxdmedia Sounds good. 

One more thing. Could you get me the info about the versions of all the various CLI tools you are using? Ionic, angular, cordova/capacitor, etc?

**Comment by jrista** [comment]
@mxdmedia Any luck with this issue?

**Comment by mxdmedia** [comment]
@jrista My apologies- Unfortunately I was sucked into the vortex of "end-of-2019/beginning-of-2020" planning stuff, and am just now getting my head back above water.  I'll take a look this afternoon/evening, and get back to you.

**Comment by mxdmedia** [comment]
For your first request- I console logged in/out of all of my components, modules, state files, etc.  In addition, I threw in some logging before and after subscribing to SampleFacade's `all$`. Here is the result:

```
[Log] AppModule: enter (user-script:2, line 78)
[Log] AppModule: post (user-script:2, line 78)
[Log] AppComponent: enter (user-script:2, line 78)
[Log] AppComponent: post (user-script:2, line 78)
[Log] routes: enter (user-script:2, line 78)
[Log] routes: post (user-script:2, line 78)
[Log] AppRoutingModule: enter (user-script:2, line 78)
[Log] AppRoutingModule: post (user-script:2, line 78)
[Log] Sample model: enter (user-script:2, line 78)
[Log] Sample model: post (user-script:2, line 78)
[Log] sample.state: enter (user-script:2, line 78)
[Log] sample.state: post buildState() (user-script:2, line 78)
[Log] sample.state: post reducer (user-script:2, line 78)
[Log] StateModule: enter (user-script:2, line 78)
[Log] StateModule: post (user-script:2, line 78)
[Log] Ionic Native: deviceready event fired after 23 ms (user-script:2, line 78)
[Log] HomePageModule: enter (user-script:2, line 78)
[Log] HomePageModule: post (user-script:2, line 78)
[Log] HomePage (component): enter (user-script:2, line 78)
[Log] Home Page (component): post (user-script:2, line 78)
[Log] SampleFacade: enter (user-script:2, line 78)
[Log] SampleFacade: post (user-script:2, line 78)
[Log] HomePage (component): ngOnInit (user-script:2, line 78)
[Log]    - Add sub to SampleFacade.all$: enter (user-script:2, line 78)
[Log]    - Add sub to SampleFacade.all$: post (user-script:2, line 78)
[Error] ERROR – TypeError: undefined is not an object (evaluating 'e.ids') — briebug-ngrx-auto-entity.js.pre-build-optimizer.js:1398
```
I did not post the error's stack trace, as it is the same as the one above.    So it appears that the error is occurring after everything has been loaded, but before the subscribe callback is called (as none of the console logging I put inside of `SampleFacade.all$.subscribe` was logged.

Regarding my cli tool versions they currently are:
Ionic: 5.4.9
Angular CLI: 8.1.3
Capacitor: 1.4.0
(not using Cordova at all)

Next I'll try updating them to latest (though my large project is using the latest of all the CLI tools, and it is showing the same errors, so I don't think that'll change anything).  I'll also try changing import order, and going without lazy loading on the home page, and report back.

**Comment by mxdmedia** [comment]
To follow up, neither changing the order if imports, nor going to static routing had any effect- the same error keeps occurring.  I think I tried just about every order of the 4 imports in my `app.module.ts` file to no avail.  Please let me know if there are any other troubleshooting steps I can take to help with resolving this.

**Comment by jrista** [comment]
@mxdmedia: I think I understand the issue now. It is interesting, most of the projects I've used NgRx Auto-Entity in so far have actually not been using prod builds. I think, some time back, I did optimized prod builds with one app, and ran into this issue, disabled optimization...and never got back to turning optimization back on. A couple others have not been using prod mode builds at all, and most of these are administrative apps of one kind of another.

I just reached the point of publishing production builds of another app I've been using Auto-Entity in, and have encountered issues I believe are related to the same underlying issue as what you are encountering. I have actually run into several issues with an Ionic 4 app, but I believe they are all related to the same thing. 

It boils down to how minification with uglifyjs, which mangles all type names, functions and class names (constructor.name). I was actually relying on constructor.name to determine the name of the state properties dynamically, and in some of my apps I created custom entity services that dynamically generated urls using the constructor.name info as well. Problem is, since the constructor.name represents the actual class name, once minified, instead of `Customer`, you might just have `e`. So your IEntityInfo.modelName becomes `e` instead of `customer`. 

For the error you are encountering, internally auto-entity is looking for a state property on AppState called, well in our example here it would be `e`, however the actual property on AppState is `customer`. So auto entity cannot actually find the state, hence the reason it is undefined, and when it tries to find undefined.ids or undefined.entities, it throws. 

We are looking into a solution. In the shorter term, it may end up being the @Entity decorator, and what we might have to do in the short term is simply use @Entity to decorate each model with a string name that is used by auto-entity. Until such time as we can either figure out how to adjust the uglifyjs settings and disable the mangling of constructor.name, or come up with another way to preserve model class names automatically. 

We'll keep you posted.

**Comment by jrista** [comment]
@mxdmedia I've implemented a required modelName property on the @Entity decorator. Along with a few other changes to make some of the code more robust and resilient in the face of missing state, I think this will solve your problem. 

I've scaled back the initial release for the @Entity decorator and hopefully it will be out soon.

**Comment by mxdmedia** [comment]
@jrista Wow, it looks like my issue sent you pretty deep into the weeds.  The whole process of uglify/minify and js code optimization is way out of my wheelhouse.  Glad you were able to find what is causing the issue, and start figuring out a way to fix it.  If you think of anything else I can do to help, I can take requests (I have a very large, non-distributable ionic-4 app that is showing the same issues.  While I cant give you access to the code, I can do some stack traces, etc..).

Just to be clear on the current fix- you are referring to the upcoming `auto-entity` entity decorator, not ngrx's entity decorator, correct?

Somewhat related- I feel bad for whomever has the github account name "entity" as they must be getting pinged like crazy from your discussions ;).

**Comment by jrista** [comment]
@mxdmedia Correct, upcoming auto-entity feature. It is actually in PR now if you want to take a look. 

I delved even deeper last night, into Angular's optimization code in the devkit. I am not sure they even use uglifyjs anymore. If they did, we could have used the webpack builder and the developer could have used the `reserved` identifiers list to keep it from mangling those identifiers (i.e. those of the model classes.) I saw nothing about uglify in the devkit code itself, and what optimizer/minimizer code I did find appeared to be 100% angular code. 

For now, we'll just have to explicitly specify thee string name of our entities with the entity decorator. I've tested it and it works with the prod build. So it will definitely solve the problem.

I also threw in the comparer feature as well, along with a .sorted$ stream on the facade base class, so you;ll be able to get your data out of state in sorted order (and memoized).

**Comment by jrista** [comment]
We are working on getting 0.4.0 out with the `@Entity` decorator. We have not yet had time to update the docs, but `@Entity` is now effectively required on all auto-entity models, and you must specify the `modelName` property in the decorator.

We are looking for a way to integrate an ng cli schematic that would automate creation of entities, and potentially whitelist their identifiers from the webpack minimizer, which would make `@Entity` optional again. We are not certain if that will be an option, as it appears the Angular team may have written their own minimizer in NG 8 and up, so UglifyJS may no longer be used. Sadly, Uglify had the `reserved` identifier names that could be configured to prevent mangling those identifiers. 

Anyway, for now...decorate all auto-entity models with `@Entity` and give them a name, at the very least. You may now also specify a sort comparer:

```typescript
@Entity({
    modelName: 'Model',
    comparer: (a: Model, b: Model): number => ((a || {}).name || '').localeCompare((b || {}).name || '')
})
export class Model {
    @Key id: number;
    name: string;
}
```

**Comment by jrista** [comment]
Resolved by #84
